contracts/concentrated/PrivatePool.sol

```diff
@@ -0,0 +1,308 @@
+        IERC20(token1).safeTransferFrom(msg.sender, address(this), amounts[1]);
+
+        uint256[] memory fees = new uint256[](2);
+        fees[0] = 0;
+        fees[1] = 0;
+
+        emit AddLiquidity(msg.sender, amounts, fees, _d, _d);
+    }
+
+    /// @notice Removes liquidity from pool
+    /// @param amount The amount of liquidity to remove
+    /// @param minAmounts The minimum amounts of token to receive in token decimals
+    /// @param deadline The deadline before which liquidity must be removed
+    function removeLiquidity(
+        uint256 amount,
+        uint256[] calldata minAmounts,
+        uint256 deadline
+    ) external onlyOwner deadlineCheck(deadline) returns (uint256[] memory amountsOut_) {
+        require(minAmounts.length == 2, "invalid min amounts");
+
+        // get current token balances in pool
+        uint256 xWad = _amountWad(IERC20(token0).balanceOf(address(this)), true);
+        uint256 yWad = _amountWad(IERC20(token1).balanceOf(address(this)), false);
+
+        // get D balance before add liquidity
+        uint256 _d = _D(xWad, yWad);
+
+        // amount of liquidity to remove must be less than D
+        require(amount <= _d, "amount > D");
+
+        // token amounts to remove are x * amount / D and y * amount / D
+        uint256 dx = _amountDecimals(Math.mulDiv(xWad, amount, _d), true);
+        uint256 dy = _amountDecimals(Math.mulDiv(yWad, amount, _d), false);
+
+        // check exceeds min amounts
+        require(dx >= minAmounts[0], "dx < min");
+        require(dy >= minAmounts[1], "dy < min");
+
+        // transfer amounts out
+        IERC20(token0).safeTransfer(msg.sender, dx);
+        IERC20(token1).safeTransfer(msg.sender, dy);
+
+        // return amounts transferred out
+        amountsOut_ = new uint256[](2);
+        amountsOut_[0] = dx;
+        amountsOut_[1] = dy;
+
+        emit RemoveLiquidity(msg.sender, amountsOut_, _d - amount);
+    }
+
+    /// @notice Swaps token from for an amount of token to
+    /// @param tokenIndexFrom The index of the token in
+    /// @param tokenIndexTo The index of the token out
+    /// @param dx The amount of token in in token decimals
+    /// @param minDy The minimum amount of token out in token decimals
+    /// @param deadline The deadline before which swap must be executed
+    function swap(
+        uint8 tokenIndexFrom,
+        uint8 tokenIndexTo,
+        uint256 dx,
+        uint256 minDy,
+        uint256 deadline
+    ) external deadlineCheck(deadline) returns (uint256 dy_) {
+        // calculate amount out from swap
+        // @dev reverts for invalid token indices
+        dy_ = calculateSwap(tokenIndexFrom, tokenIndexTo, dx);
+        require(dy_ >= minDy, "dy < minDy");
+
+        // transfer dx in and send dy out
+        address tokenIn = tokenIndexFrom == 0 ? token0 : token1;
+        address tokenOut = tokenIndexTo == 0 ? token0 : token1;
+        IERC20(tokenIn).safeTransferFrom(msg.sender, address(this), dx);
```

Comment:
To be on the safe side I would consider using the pattern from `Swap.sol`, first transfer tokens, and use the balance increment as `dx`. This helps in the unlikely case that a pooled token will have the fee switched on (for instance, `USDT` has the fee switch, so the best assurance from Tether is that they pinky swore they won't turn it on).

https://github.com/synapsecns/synapse-contracts/blob/b576a6b89a4d2382c82eead74083a41c9db7a538/contracts/amm/SwapUtils.sol#L682-L694

Reply:
Updated!